checkpoint/keeper_milestone.go

```diff
@@ -0,0 +1,208 @@
+package checkpoint
+
+import (
+	"errors"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	cmn "github.com/maticnetwork/heimdall/common"
+	hmTypes "github.com/maticnetwork/heimdall/types"
+)
+
+var (
+	MilestoneKey          = []byte{0x20} // Key to store milestone
+	CountKey              = []byte{0x30} //Key to store the count
+	MilestoneNoAckKey     = []byte{0x40} //Key to store the NoAckMilestone
+	MilestoneLastNoAckKey = []byte{0x50} //Key to store the Latest NoAckMilestone
+	LastMilestoneTimeout  = []byte{0x60} //Key to store the Last Milestone Timeout
+)
+
+// AddMilestone adds milestone into final blocks
+func (k *Keeper) AddMilestone(ctx sdk.Context, milestone hmTypes.Milestone) error {
+	milestoneNumber := k.GetMilestoneCount(ctx) + 1 //GetCount gives the number of previous milestone
+
+	key := GetMilestoneKey(milestoneNumber)
+	if err := k.addMilestone(ctx, key, milestone); err != nil {
+		return err
+	}
+
+	pruningNumber := milestoneNumber - 100
+
+	k.PruneMilestone(ctx, pruningNumber) //Prune the old milestone to reduce the memory consumption
+	k.SetMilestoneCount(ctx, milestoneNumber)
+	k.Logger(ctx).Info("Adding good milestone to state", "milestone", milestone, "milestoneNumber", milestoneNumber)
+
+	return nil
+}
+
+// addMilestone adds milestone to store
+func (k *Keeper) addMilestone(ctx sdk.Context, key []byte, milestone hmTypes.Milestone) error {
+	store := ctx.KVStore(k.storeKey)
+
+	// create Checkpoint block and marshall
+	out, err := k.cdc.MarshalBinaryBare(milestone)
+	if err != nil {
+		k.Logger(ctx).Error("Error marshalling milestone", "error", err)
+		return err
+	}
+
+	// store in key provided
+	store.Set(key, out)
+
+	return nil
+}
+
+// GetMilestoneKey appends prefix to milestoneNumber
+func GetMilestoneKey(milestoneNumber uint64) []byte {
+	milestoneNumberBytes := []byte(strconv.FormatUint(milestoneNumber, 10))
+	return append(MilestoneKey, milestoneNumberBytes...)
+}
+
+// GetMilestoneByNumber to get milestone by milestone number
+func (k *Keeper) GetMilestoneByNumber(ctx sdk.Context, number uint64) (*hmTypes.Milestone, error) {
+	store := ctx.KVStore(k.storeKey)
+	milestoneKey := GetMilestoneKey(number)
+
+	var _milestone hmTypes.Milestone
+
+	if store.Has(milestoneKey) {
+		if err := k.cdc.UnmarshalBinaryBare(store.Get(milestoneKey), &_milestone); err != nil {
+			return nil, err
+		}
+
+		return &_milestone, nil
+	}
+
+	return nil, errors.New("Invalid milestone Index")
+}
+
+// GetLastMilestone gets last milestone, milestone number = GetCount()
+func (k *Keeper) GetLastMilestone(ctx sdk.Context) (*hmTypes.Milestone, error) {
```

Comment:
A lot of code duplication. This is ok at the moment, but could you create a new task to remove code duplication from the keeper?

Reply:
I am about the most of the content of the function. These fetch* are quite the same.

Reply:
Yes, I think we need to make new task for it as all our keeper functions in all the module are very similar in content.